### PR TITLE
[SPARK-3767] [CORE] Support wildcard in Spark properties

### PR DESCRIPTION
**Pull request by devaraj-kavali**
## What changes were proposed in this pull request?

Added provision to specify the 'spark.executor.extraJavaOptions' value in terms of the Executor Id(i.e. @execid@). @execid@ will be replaced with the Executor Id while starting the executor.
## How was this patch tested?

I have verified this by checking the executor process command  and gc logs. I verified the same in different deployement modes(Standalone, YARN, Mesos).

### CODE REVIEW

**Comment by rxin** [comment]
I think this is a good idea in general, but I'm less sure about @. Is @ used in windows? Any other places?

**Comment by devaraj-kavali** [comment]
Thanks @rxin for checking this, I don't think @ is used any where. Here again we are replacing only for 'spark.executor.extraJavaOptions' value when @execid@ occurs, any other @ symbols we leave as it is, so I don't think any problem occurs due to this.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 1]
```diff
@@ -166,14 +166,15 @@ private[spark] class CoarseMesosSchedulerBackend(
       environment.addVariables(
         Environment.Variable.newBuilder().setName("SPARK_CLASSPATH").setValue(cp).build())
     }
-    val extraJavaOpts = conf.get("spark.executor.extraJavaOptions", "")
+    var extraJavaOpts = conf.get("spark.executor.extraJavaOptions", "")
```
Comment (BryanCutler) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: would it be a little cleaner to do the following
```scala
val extraJavaOpts = conf.getOption("spark.executor.extraJavaOptions").map {
  Utils.substituteExecIdWildCard(_, taskId)
}.getOrElse("")
```

Comment (devaraj-kavali) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: Thanks @BryanCutler  for the suggestion, I have addressed it in the latest.

**Comment by devaraj-kavali** [comment]
@rxin, Please have a look into this and let me know any thing needs to be done here. About @, M/R also uses @ for the taskid wild card in java opts and there is no problem in windows and as well as in other places with @.

**Comment by vanzin** [comment]
Sorry this has been forgotten for so long. I think instead of adding a new way of referencing variables, it might be better to use the new code I'm adding in #14468; that way it would be easier to reference executor id (by creating a new `ConfigProvider` with that info) and even other config information if desired.

At the very least conflicts need to be resolved.

**Comment by devaraj-kavali** [comment]
@vanzin Thanks for looking into this, I have resolved the conflicts.

**Comment by vanzin** [comment]
Hmmm... wouldn't this be a one line change (for standalone and mesos) if instead you modified `CommandUtils.buildLocalCommand` to apply variable substitution to `command.javaOpts` instead of just to `command.arguments`?

That would replace `{{EXECUTOR_ID}}` in any of the arguments, since that variable is already in the list of things to replace (see `ExecutorRunner.substituteVariables`.)

You'd still need some extra code for YARN since it doesn't use that path. But it would be a smaller change that re-uses existing functionality.

**Comment by vanzin** [comment]
Also the PR title is misleading, since this change is adding a single "wildcard" to a single configuration. The title makes it sound like it's way more generic than it is.

**Comment by vanzin** [comment]
Hi @devaraj-kavali, are you still interested in updating this PR? I really think it should use the features I added in SPARK-16671 (especially the `ConfigReader` code), to avoid yet another way of expanding variables.

**Comment by devaraj-kavali** [comment]
@vanzin, SPARK-3767 was resolved as 'Won't Fix' by @srowen. I was in assumption that SPARK-16671 covers this as well.

**Comment by vanzin** [comment]
If you're willing to work on it you can re-open the bug. Otherwise you should close this PR.

**Comment by devaraj-kavali** [comment]
I will update this PR with the ConfigReader and reopen the jira.

**Pull request closed without merging**